testing/internal/e2e/tests/base/target_pagination_test.go

```diff
@@ -0,0 +1,181 @@
+			"-format=json",
+		),
+	)
+	require.NoError(t, output.Err, string(output.Stderr))
+
+	var newTargets targets.TargetListResult
+	err = json.Unmarshal(output.Stdout, &newTargets)
+	require.NoError(t, err)
+
+	require.Len(t, initialTargets.Items, globals.DefaultMaxPageSize+1)
+	assert.Equal(t, newTargetId, newTargets.Items[len(newTargets.Items)-1].Id)
+	assert.Empty(t, initialTargets.ResponseType)
+	assert.Empty(t, initialTargets.RemovedIds)
+	assert.Empty(t, initialTargets.RefreshToken)
+	for _, target := range newTargets.Items {
+		assert.NotEqual(t, target.Id, initialTargets.Items[0].Id)
+	}
+}
+
+// TestApiPaginateTargets asserts that the API automatically paginates to retrieve
+// all targets in a single invocation.
+func TestApiPaginateTargets(t *testing.T) {
+	e2e.MaybeSkipTest(t)
+	c, err := loadTestConfig()
+	require.NoError(t, err)
+
+	client, err := boundary.NewApiClient()
+	require.NoError(t, err)
+	ctx := context.Background()
+	sClient := scopes.NewClient(client)
+	tClient := targets.NewClient(client)
+	newOrgId := boundary.CreateNewOrgApi(t, ctx, client)
+	t.Cleanup(func() {
+		ctx := context.Background()
+		_, err := sClient.Delete(ctx, newOrgId)
+		require.NoError(t, err)
+	})
+	newProjectId := boundary.CreateNewProjectApi(t, ctx, client, newOrgId)
+
+	// Create enough targets to overflow a single page.
+	targetPort, err := strconv.ParseInt(c.TargetPort, 10, 32)
+	require.NoError(t, err)
+	for i := 0; i < globals.DefaultMaxPageSize+1; i++ {
+		_, err := tClient.Create(ctx, "tcp", newProjectId,
+			targets.WithName("test-target-"+strconv.Itoa(i)),
+			targets.WithTcpTargetDefaultPort(uint32(targetPort)),
+			targets.WithAddress(c.TargetAddress),
+		)
+		require.NoError(t, err)
+	}
+
+	// List targets recursively
+	initialTargets, err := tClient.List(ctx, newProjectId)
+	require.NoError(t, err)
+
+	require.Len(t, initialTargets.Items, globals.DefaultMaxPageSize+1)
+	assert.Equal(t, "complete", initialTargets.ResponseType)
+	assert.Empty(t, initialTargets.RemovedIds)
+	assert.NotEmpty(t, initialTargets.RefreshToken)
+	mapItems, ok := initialTargets.GetResponse().Map["items"]
+	require.True(t, ok)
+	mapSliceItems, ok := mapItems.([]any)
+	require.True(t, ok)
+	assert.Len(t, mapSliceItems, globals.DefaultMaxPageSize+1)
+
+	// Create a new target and destroy one of the other targets
+	newTargetResult, err := tClient.Create(ctx, "tcp", newProjectId,
+		targets.WithName("test-target-"+strconv.Itoa(globals.DefaultMaxPageSize+1)),
+		targets.WithTcpTargetDefaultPort(uint32(targetPort)),
+		targets.WithAddress(c.TargetAddress),
+	)
+	require.NoError(t, err)
+	_, err = tClient.Delete(ctx, initialTargets.Items[0].Id)
+	require.NoError(t, err)
+
+	// List again, should have the new and deleted target
+	newTargets, err := tClient.List(ctx, newProjectId, targets.WithRefreshToken(initialTargets.RefreshToken))
+	require.NoError(t, err)
+
+	require.Len(t, newTargets.Items, 1)
+	assert.Equal(t, newTargetResult.Item.Id, newTargets.Items[0].Id)
```

Comment:
Is this always going to be the case? The new target id is going to the first element? Also, is that different from the CLI? It looks like we're checking that the new target is the last in the list?

Reply:
Yes and yes. It should always be the first (and only) element returned, since we're using a refresh token created by the previous call. Only new targets and updated targets since that List call will show up in this call. Since this is the only new or updated target, it's the only one to show up. I could add a target update as well to ensure that shows up here 👍🏻 .

Reply:
It is also different from the CLI because the CLI doesn't expose any of the pagination logic to the user (at the moment). There's no flag to provide a refresh token, so there's little point in returning it in the result.